### PR TITLE
feat(griffin): distinct queries on symbol column retrieve the list fr…

### PR DESCRIPTION
**Pull request by jaugsburger**
…om metadata instead of iterating through the table.

### CODE REVIEW

File: core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java [hunk 1]
```diff
@@ -1110,6 +1110,21 @@ private RecordCursorFactory generateSelectChoose(QueryModel model, SqlExecutionC
     private RecordCursorFactory generateSelectDistinct(QueryModel model, SqlExecutionContext executionContext) throws SqlException {
         final RecordCursorFactory factory = generateSubQuery(model, executionContext);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java]: this factory does not need creating if condition below is met

File: core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java [hunk 2]
```diff
@@ -1110,6 +1110,21 @@ private RecordCursorFactory generateSelectChoose(QueryModel model, SqlExecutionC
...
         try {
+            if (model.getColumns().size() == 1 && model.getNestedModel() != null && model.getNestedModel().getNestedModel() != null && model.getNestedModel().getNestedModel().getTableName() != null) {
+                ExpressionNode tableName = model.getNestedModel().getNestedModel().getTableName();
+                TableReader reader = engine.getReader(executionContext.getCairoSecurityContext(), tableName.token);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java]: reader is a shared resource - SQL code generator must not assign it to the factory, but close it instead in try() block.

File: core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java [hunk 3]
```diff
@@ -1110,6 +1110,21 @@ private RecordCursorFactory generateSelectChoose(QueryModel model, SqlExecutionC
...
+                CharSequence columnName = model.getColumnNames().get(0);
+                int columnIndex = reader.getMetadata().getColumnIndex(columnName);
+                int columnType = reader.getMetadata().getColumnType(columnIndex);
+                if (columnType == ColumnType.SYMBOL) {
+                    return new DistinctSymbolRecordCursorFactory(
+                            factory,
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java]: this cursor should not need base factory

File: core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java [hunk 4]
```diff
@@ -1110,6 +1110,21 @@ private RecordCursorFactory generateSelectChoose(QueryModel model, SqlExecutionC
...
+                            entityColumnFilter,
+                            reader,
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java]: Reader must not be held by factory. This is a resource hungry instance, which are shared through the engine. Factory must hold on to the CairoEngine instance instead and table name.

File: core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java [hunk 5]
```diff
@@ -1110,6 +1110,21 @@ private RecordCursorFactory generateSelectChoose(QueryModel model, SqlExecutionC
...
+                            columnName
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java]: it is better to pass column index

File: core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+public class DistinctSymbolRecordCursorFactory implements RecordCursorFactory {
+    private final DistinctSymbolRecordCursor cursor;
+    private final RecordMetadata metadata;
+    private final TableReader reader;
+    private final CharSequence columnName;
+
+    public DistinctSymbolRecordCursorFactory(
+            RecordCursorFactory base,
+            EntityColumnFilter columnFilter,
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java]: not sure why we need column filter?

File: core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+        this.columnName = columnName;
+    }
+
+    @Override
+    public void close() {
+    }
+
+    @Override
+    public RecordCursor getCursor(SqlExecutionContext executionContext) {
+        cursor.of(reader, columnName);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java]: when you have CairoEngine - this is the place to get reader

File: core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+        private TableReader reader;
+        private SymbolMapReader symbolMapReader;
+        private int numberOfSymbols;
+
+        DistinctSymbolRecordCursor() {
+        }
+
+        @Override
+        public void close() {
+            if (reader != null) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java]: there is method Misc.free() that does something like this

File: core/src/test/java/io/questdb/griffin/SqlCodeGeneratorTest.java [hunk 9]
```diff
@@ -1104,6 +1104,31 @@ public void testSumDoubleColumn() throws Exception {
         }
     }
 
+
+    @Test
+    public void testDistinctSymbolColumn() throws Exception {
```
Comment (bluestreak01) [inline core/src/test/java/io/questdb/griffin/SqlCodeGeneratorTest.java]: we need one more test where symbol value is NULL

File: core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+        @Override
+        public void toTop() {
+            recordA.recordIndex = -1;
+        }
+
+        public void of(TableReader reader, int columnIndex) {
+            this.symbolMapReader = reader.getSymbolMapReader(columnIndex);
+            this.numberOfSymbols = symbolMapReader.size();
+            this.reader = reader;
+            this.recordA.of(reader, columnIndex);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java]: you shouldn't need to pass TableReader to record. Record needs SymbolMapReader

File: core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+
+    public static class DistinctSymbolRecord implements Record {
+        private int columnIndex;
+        private int recordIndex = -1;
+        private TableReader reader;
+
+        @Override
+        public CharSequence getSym(int col) {
+            if (reader != null) {
+                return reader.getSymbolMapReader(columnIndex).valueOf(recordIndex);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java]: we can avoid calling for every value of the result set

```reader.getSymbolMapReader(columnIndex)```

File: core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    public static class DistinctSymbolRecord implements Record {
+        private int columnIndex;
+        private int recordIndex = -1;
+        private TableReader reader;
+
+        @Override
+        public CharSequence getSym(int col) {
+            if (reader != null) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java]: reader null check should not be necessary when Record is setup correctly. If reader is every to become null this should be reported where cursor is requested (don't think null is possible at all in this case)

File: core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    @Override
+    public boolean isRandomAccessCursor() {
+        return true;
+    }
+
+    private static class DistinctSymbolRecordCursor implements RecordCursor {
+        private DistinctSymbolRecord recordA = new DistinctSymbolRecord();
+        private DistinctSymbolRecord recordB = new DistinctSymbolRecord();
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java]: I would suggest making `DistinctSymbolRecord` non-static class. This way it can access symbol map reader on factory instance and only have its own record index.

This said, `recordB` is better be created lazily. It is only used when someone is ordering symbols returned by this cursor. Which is not very often.

File: core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java [hunk 14]
```diff
@@ -1108,6 +1108,37 @@ private RecordCursorFactory generateSelectChoose(QueryModel model, SqlExecutionC
...
+        if (model.getColumns().size() == 1 && model.getNestedModel() != null && model.getNestedModel().getNestedModel() != null && model.getNestedModel().getNestedModel().getTableName() != null) {
+            ExpressionNode tableNameExpressionNode = model.getNestedModel().getNestedModel().getTableName();
+            CharSequence tableName = tableNameExpressionNode.token;
+            TableReader reader = engine.getReader(executionContext.getCairoSecurityContext(), tableName);
+            CharSequence columnName = model.getColumnNames().get(0);
+            RecordMetadata readerMetadata = reader.getMetadata();
+            int columnIndex = readerMetadata.getColumnIndex(columnName);
+            int columnType = readerMetadata.getColumnType(columnIndex);
+            if (columnType == ColumnType.SYMBOL) {
+                final GenericRecordMetadata disticntSymbolMetadata = new GenericRecordMetadata();
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java]: spelling `disticntSymbolMetadata `

File: core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+        private TableReader reader;
+        private SymbolMapReader symbolMapReader;
+        private int numberOfSymbols;
+
+        DistinctSymbolRecordCursor() {
+        }
+
+        @Override
+        public void close() {
+            Misc.free(reader);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java]: it would be better to assign `reader`, e.g.

```
reader = Misc.free(reader);
```

to defend against double close

File: core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+        this.metadata = metadata;
+        this.tableName = tableName;
+        this.columnIndex = columnIndex;
+        this.tableVersion = tableVersion;
+        this.cursor = new DistinctSymbolRecordCursor();
+    }
+
+    @Override
+    public void close() {
+    }
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java]: factory must close its cursor (which is closeable)

File: core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+        }
+
+        @Override
+        public void recordAt(Record record, long atRowId) {
+            ((DistinctSymbolRecord) record).recordIndex = (int) atRowId;
+        }
+
+        @Override
+        public void toTop() {
+            recordA.recordIndex = -1;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java]: minor: toTop() and record.reset() implement the same logic, perhaps `topTop()` can call `recordA.reset()`. Or may be `record.reset()` could be renamed to `record.toTop()`

File: core/src/main/java/io/questdb/cairo/SymbolMapReader.java [hunk 18]
```diff
@@ -35,4 +35,6 @@
     boolean isDeleted();
 
     void updateSymbolCount(int count);
+
+    boolean containsNullValue();
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/SymbolMapReader.java]: Can we pull this value up `StaticSymbolTable` interface level? The idea of static symbol table is that it is bounded, e.g. knows of its size and presumably that it knows that it has null?

Comment (jaugsburger) [reply core/src/main/java/io/questdb/cairo/SymbolMapReader.java]: sure

File: core/src/main/java/io/questdb/cairo/SymbolMapWriter.java [hunk 19]
```diff
@@ -155,6 +158,7 @@ public int getSymbolCount() {
     public int put(CharSequence symbol) {
 
         if (symbol == null) {
+            offsetMem.putBool(HEADER_NULL_FLAG, true);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/SymbolMapWriter.java]: this write should be cached i feel like, e.g. read of class member is faster than write to memory mapped file

File: core/src/test/java/io/questdb/griffin/AbstractGriffinTest.java [hunk 20]
```diff
@@ -290,7 +287,11 @@ private static void testSymbolAPI(RecordMetadata metadata, RecordCursor cursor)
                     if (symbolTable instanceof StaticSymbolTable) {
                         CharSequence sym = record.getSym(column);
                         int value = record.getInt(column);
-                        Assert.assertEquals(value, ((StaticSymbolTable) symbolTable).keyOf(sym));
+                        if (symbolTable instanceof SymbolMapReader && ((SymbolMapReader) symbolTable).containsNullValue() && value == ((StaticSymbolTable) symbolTable).size()) {
```
Comment (bluestreak01) [inline core/src/test/java/io/questdb/griffin/AbstractGriffinTest.java]: I don't follow this condition. Why are we comparing int value of symbol to size of symbol map?

File: core/src/main/java/io/questdb/cairo/SymbolMapWriter.java (hunk 19 above)
Comment (jaugsburger) [reply core/src/main/java/io/questdb/cairo/SymbolMapWriter.java]: yes, working on that, just wanted to push the latest changes

File: core/src/main/java/io/questdb/cairo/SymbolMapReader.java (hunk 18 above)
Comment (bluestreak01) [reply core/src/main/java/io/questdb/cairo/SymbolMapReader.java]: 👍

File: core/src/test/java/io/questdb/griffin/AbstractGriffinTest.java (hunk 20 above)
Comment (jaugsburger) [reply core/src/test/java/io/questdb/griffin/AbstractGriffinTest.java]: the reason is that the symboMapReader.size() still returns the number of symbols without counting the null value. on the other hand, record.hasNext() now returns true one final time if the symbolMapReader contains a null symbol

Comment (bluestreak01) [reply core/src/test/java/io/questdb/griffin/AbstractGriffinTest.java]: perfect, understood, thanks!

File: core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+        public class DistinctSymbolRecord implements Record {
+            private int recordIndex = -1;
+
+            public void decrementRecordIndex() {
+                recordIndex--;
+            }
+
+            @Override
+            public CharSequence getSym(int col) {
+                return reader.getSymbolMapReader(columnIndex).valueOf(recordIndex);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java]: missed a spot
